Add WebsiteReindexProviderVisitor to extract indexable content for article/page

Prokyonn opened the pull request:
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| License | MIT

#### What's in this PR?

- Adds a visitor interface to the WebsitePageReindexProvider and WebsiteArticleReindexProvider
- Adds a visitor that extracts the indexable data from the content for both entities

#### Why?

For website search the content should also be indexed.

Prokyonn committed:
Add WebsitePageReindexProviderVisitor

Prokyonn committed:
Add same visitor pattern for articles

alexander-schranz commented on packages/page/src/Infrastructure/Sulu/Search/WebsitePageReindexProvider.php:
thought we returned here a Page and could give the Page Entity to the visitor 🤔 but seems like false marked.

Prokyonn replied:
Yes unfortunately it is only the Page `phpstan-type`, I guess this was implemented without hydration to improve performance

Prokyonn replied:
but we could also add a visitor to the query builder, than it would still be extendable also for custom adjustments on the page entity

alexander-schranz commented on packages/article/src/Infrastructure/Sulu/Search/Visitor/WebsiteArticleReindexContentVisitor.php:
```diff
@@ -0,0 +1,225 @@
+
+            $hasSearchTag = false;
+            foreach ($item->getTags() as $tag) {
+                if ('sulu.search.field' === $tag->getName()) {
```
Can we make a constant for this here.

alexander-schranz commented on packages/page/src/Infrastructure/Sulu/Search/Visitor/WebsitePageReindexProviderVisitorInterface.php:
```diff
@@ -0,0 +1,31 @@
+     *
+     * @return array<string, mixed>
+     */
+    public function visit(array $result, array $data): array;
```
```suggestion
    public function enhanceDocument(array $queryResult, array $document): array;
    
    public function enhanceQuery(QueryBuilder $queryBuilder): void;
```

Prokyonn replied:
Sounds good 👍

alexander-schranz commented on packages/page/src/Infrastructure/Sulu/Search/Visitor/WebsitePageReindexProviderVisitorInterface.php:
```diff
@@ -0,0 +1,31 @@
+ * Interface for visitors that can process and manipulate reindex data
+ * for website page dimension content during search indexing.
+ */
+interface WebsitePageReindexProviderVisitorInterface
```
```suggestion
interface WebsitePageReindexProviderEnhancerInterface
```

alexander-schranz reviewed (commented):
Did discuss with Thomas that maybe this make sense.

What do you think about?

Prokyonn committed:
Refactor to enhancer

# Conflicts:
#	packages/page/src/Infrastructure/Sulu/Search/WebsitePageReindexProvider.php

alexander-schranz commented:
@Prokyonn looks good 👍 Rebase required.
